libvast/native-plugins/rebuild.cpp

```diff
@@ -38,6 +41,196 @@ namespace vast::plugins::rebuild {
+    VAST_ASSERT(resize_result.ok(), resize_result.ToString().c_str());
+  }
+
+  caf::error
+  add(type schema, std::shared_ptr<arrow::RecordBatch> batch) override {
+    VAST_ASSERT(schema == schema_);
+    VAST_ASSERT(builder_ != nullptr);
+    const auto batch_size = batch->num_rows();
+    if (builder_->length() == 0 && batch_size == desired_batch_size_) {
+      result_.push_back(batch);
+      return caf::none;
+    }
+    for (auto offset = int64_t{}; offset < batch->num_rows();) {
+      const auto length = std::min(desired_batch_size_ - builder_->length(),
+                                   batch_size - offset);
+      append_slice_to_builder(batch, offset, length);
+      offset += length;
+      VAST_ASSERT(builder_->length() <= desired_batch_size_);
+      if (builder_->length() == desired_batch_size_)
+        finish_builder();
+    }
+    return caf::none;
+  }
+
+  caf::expected<std::vector<pipeline_batch>> finish() override {
+    if (builder_->length() > 0)
+      finish_builder();
+    auto result = std::vector<pipeline_batch>{};
+    result.reserve(result_.size());
+    for (auto&& batch : std::exchange(result_, {}))
+      result.emplace_back(schema_, std::move(batch));
+    return result;
+  }
+
+private:
+  /// Appends a slice of a batch to the currently ongoing builder.
+  /// @param batch The record batch to partially rebuild.
+  /// @param offset The starting row inside the batch.
+  /// @param length THe length of the slice for our batch.
+  ///
+  /// @note You'd think we could call into Arrow here, but the underlying
+  /// functionality for merging arrays is neither generically implemented for
+  /// extension types, nor is it extensible by implementers of extension types.
+  /// If it was implemented, then this is how that would look like. We last
+  /// tested this against Arrow 9.0.0.
+  ///
+  ///   const auto array = batch->ToStructArray().ValueOrDie();
+  ///   const auto append_result = builder_->AppendArraySlice(
+  ///     *array->data(), offset, length);
+  ///   VAST_ASSERT(append_result.ok(), append_result.ToString().c_str());
+  ///
+  /// So what we do instead here is that we simply append the columns manually.
+  /// We don't go through the table slice builder API as that is a row-major
+  /// API and we want a column-major builder here.
+  void append_slice_to_builder(const std::shared_ptr<arrow::RecordBatch>& batch,
+                               int64_t offset, int64_t length) noexcept {
+    const auto append_columns
+      = [](const auto& self, const record_type& schema,
+           const type_to_arrow_array_t<record_type>& array,
+           type_to_arrow_builder_t<record_type>& builder, int64_t offset,
+           int64_t length) noexcept -> void {
+      // Ideally we'd assert here that the array length is at most offset +
+      // length, but the array length may actually be less than that, because
+      //
+      // Arrow silently shortens nested Arrays with nulls at the end.
```

Comment:
```suggestion
      // Ideally we'd assert here that the array length is at least offset +
      // length, but the array length may actually be less than that, because
      // Arrow silently shortens nested Arrays with nulls at the end.
      // 
```